### ISSUE TITLE
Chapter 10: Bug in definition of foldLeft?

### ISSUE THREAD

**Issue by chrdebru** [fpinscala/fpinscala#667]
The book creates an extension for `List[A]`, and thus Scala relies on the implementations of `foldRight`, `foldLeft`, and `toList` of `List[A]`. 

Only the definition of `foldMap` will call the definition of `foldRight` of that given. So the definitions of foldLeft seem to behave nicely. I am, however, unable to comprehend how the definition of `foldLeft` can work.  I tried manually using the substation model and was stumped by the use of dual of endoMonoid. I believe there may be an error.

Using a very minimal custom List ADT.

```scala
package fpinscala.answers.datastructures

enum List[+A]:
  case Nil
  case Cons(head: A, tail: List[A])

object List: 
  def apply[A](as: A*): List[A] = 
    if as.isEmpty then Nil
    else Cons(as.head, apply(as.tail*))

  def append[A](a1: List[A], a2: List[A]): List[A] =
    a1 match
      case Nil => a2
      case Cons(h,t) => Cons(h, append(t, a2))
```

To see which functions are called in what order, I used:

```scala
import fpinscala.answers.monoids.Foldable.{given, *}
import fpinscala.answers.datastructures.List._
import fpinscala.answers.datastructures.List

def main(args: Array[String]): Unit = 
    val l = List(1,2,3,4,5)
    val f = (a: Int, b: Int) => {println(s"a = $a, b = $b") ; a + b}
    println(l.foldRight(0)(f))
    println(l.foldLeft(0)(f))
```

The output is
```
a = 5, b = 0
a = 4, b = 5
a = 3, b = 9
a = 2, b = 12
a = 1, b = 14
15
a = 0, b = 5
a = 5, b = 4
a = 9, b = 3
a = 12, b = 2
a = 14, b = 1
15
```

So `foldLeft` does not start with 1, but with 5. The desired behavior is achieved when relying on `foldLeft` of Scala's `List`. When manually tracing, the reversion of `f andThen g` seems to mix it up. Is it possible that dual should be omitted in that definition?

**Comment by chrdebru** [comment]
Apologies, here's the given:

```scala
  import fpinscala.answers.datastructures.List
  import fpinscala.answers.datastructures.List._

  given Foldable[List] with
    extension [A](as: List[A])
      override def foldRight[B](acc: B)(f: (A, B) => B) = as match
        case Nil => acc
        case Cons(x, xs) => f(x, xs.foldRight(acc)(f))
```

**Comment by chrdebru** [comment]
Upon further reflection, the definition of `foldRight` should use the dual and not `foldLeft`. Could it be that endoMonoid used to be defined in terms of `compose`?

**Comment by mpilquist** [comment]
Thanks for reporting. Yes, `foldRight` should be defined using the dual -- the manuscript even walks through defining `foldRight` with the dual and `foldLeft` without. Corrected in 4aab14f1e33da401481e36ab8e34e2a65845a216.
